File: submission/nikhil-funde/src/test/java/com/kulu/sdet/support/TransferApiClient.java
package com.kulu.sdet.support;

import static io.restassured.RestAssured.given;

import com.kulu.sdet.service.model.TransferRequest;
import io.restassured.http.ContentType;
import io.restassured.response.Response;
import java.util.LinkedHashMap;
import java.util.Map;

public class TransferApiClient {

  public Response createTransfer(TransferRequest request) {
    return createTransfer(request, null);
  }

  public Response createTransfer(TransferRequest request, String idempotencyKey) {
    var spec = given().contentType(ContentType.JSON).body(toMap(request));
    if (idempotencyKey != null) {
      spec = spec.header("Idempotency-Key", idempotencyKey);
    }
    return spec.when().post("/transfers");
  }

  public Response createTransfer(Map<String, Object> body) {
    return given().contentType(ContentType.JSON).body(body).when().post("/transfers");
  }

  public Response createTransfer(Map<String, Object> body, String idempotencyKey) {
    var spec = given().contentType(ContentType.JSON).body(body);
    if (idempotencyKey != null) {
      spec = spec.header("Idempotency-Key", idempotencyKey);
    }
    return spec.when().post("/transfers");
  }

  public Response getTransfer(String transferId) {
    return given().when().get("/transfers/{transferId}", transferId);
  }

  public Response getWallet(String walletId) {
    return given().when().get("/wallets/{walletId}", walletId);
  }

  private Map<String, Object> toMap(TransferRequest request) {
    Map<String, Object> map = new LinkedHashMap<>();
    if (request.sourceWalletId != null) {
      map.put("source_wallet_id", request.sourceWalletId);
    }
    if (request.destinationWalletId != null) {
      map.put("destination_wallet_id", request.destinationWalletId);
    }
    if (request.amount != null) {
      map.put("amount", request.amount);
    }
    if (request.currency != null) {
      map.put("currency", request.currency);
    }
    if (request.reference != null) {
      map.put("reference", request.reference);
    }
    return map;
  }
}

File: submission/solution-parimal-jagtap/helpers/transfer-api-client.ts
import { APIRequestContext } from '@playwright/test';

export interface TransferPayload {
  source_wallet_id: string;
  destination_wallet_id: string;
  amount: number;
  currency: string;
  reference: string;
}

export interface TransferResponse {
  transfer_id: string;
  source_wallet_id: string;
  destination_wallet_id: string;
  amount: number;
  currency: string;
  reference: string;
  status: 'COMPLETED' | 'FAILED' | 'PENDING';
  idempotency_key: string | null;
  created_at: string;
  updated_at: string;
}

export interface WalletResponse {
  wallet_id: string;
  balance: number;
  currency: string;
  created_at: string;
}

export class TransferApiClient {
  constructor(private request: APIRequestContext) {}

  async createTransfer(payload: TransferPayload, idempotencyKey?: string) {
    const headers: Record<string, string> = {};
    if (idempotencyKey) {
      headers['Idempotency-Key'] = idempotencyKey;
    }
    return this.request.post('/transfers', { data: payload, headers });
  }

  async getTransfer(transferId: string) {
    return this.request.get(`/transfers/${transferId}`);
  }

  async getWallet(walletId: string) {
    return this.request.get(`/wallets/${walletId}`);
  }

  // Test-only: reset DB between tests
  async resetDatabase() {
    return this.request.post('/test/reset');
  }
}
